notify/jira/jira.go

```diff
@@ -0,0 +1,339 @@
+
+	requestBody := issueSearch{}
+	requestBody.Jql = jql.String()
+	requestBody.MaxResults = 2
+	requestBody.Fields = []string{"status"}
+	requestBody.Expand = []string{}
+
+	level.Debug(n.logger).Log("msg", "search for recent issues", "alert", key.String(), "jql", jql.String())
+
+	responseBody, shouldRetry, err := n.doAPIRequest(tmplTextFunc, http.MethodPost, "search", requestBody)
+	if err != nil {
+		return nil, shouldRetry, err
+	}
+
+	var issueSearchResult issueSearchResult
+	err = json.Unmarshal(responseBody, &issueSearchResult)
+	if err != nil {
+		return nil, false, err
+	}
+
+	switch issueSearchResult.Total {
+	case 0:
+		level.Debug(n.logger).Log("msg", "found no existing issue", "alert", key.String())
+		return nil, false, nil
+	default:
+		level.Warn(n.logger).Log("msg", "more than one issue matched, picking most recently resolved", "alert", key.String(), "picked", issueSearchResult.Issues[0].Key)
+
+		fallthrough
+	case 1:
+		return &issueSearchResult.Issues[0], false, nil
+	}
+}
+
+func (n *Notifier) getIssueTransitionByName(tmplTextFunc templateFunc, issueKey, transitionName string) (string, bool, error) {
+	path := fmt.Sprintf("issue/%s/transitions", issueKey)
+
+	responseBody, shouldRetry, err := n.doAPIRequest(tmplTextFunc, http.MethodGet, path, nil)
+	if err != nil {
+		return "", shouldRetry, err
+	}
+
+	var issueTransitions issueTransitions
+	err = json.Unmarshal(responseBody, &issueTransitions)
+	if err != nil {
+		return "", false, err
+	}
+
+	for _, issueTransition := range issueTransitions.Transitions {
+		if issueTransition.Name == transitionName {
+			return issueTransition.ID, false, nil
+		}
+	}
+
+	return "", false, fmt.Errorf("can't find transition %s for issue %s", transitionName, issueKey)
+}
+
+func (n *Notifier) TransitionIssue(tmplText templateFunc, key notify.Key, issueKey, transitionName string) (bool, error) {
+	transitionID, shouldRetry, err := n.getIssueTransitionByName(tmplText, issueKey, transitionName)
+	if err != nil {
+		return shouldRetry, err
+	}
+
+	requestBody := issue{}
+	requestBody.Transition = &idNameValue{ID: transitionID}
+
+	path := fmt.Sprintf("issue/%s/transitions", issueKey)
+
+	level.Debug(n.logger).Log("msg", "transitions jira issue", "alert", key.String(), "key", issueKey, "transition", transitionName)
+	_, shouldRetry, err = n.doAPIRequest(tmplText, http.MethodPost, path, requestBody)
+
+	return shouldRetry, err
+}
+
+func (n *Notifier) doAPIRequest(tmplTextFunc templateFunc, method, path string, requestBody any) ([]byte, bool, error) {
+	var body io.Reader
+	if requestBody != nil {
+		var buf bytes.Buffer
+		if err := json.NewEncoder(&buf).Encode(requestBody); err != nil {
+			return nil, false, err
+		}
+
+		body = &buf
+	}
+
+	url := n.conf.APIURL.JoinPath(path)
+	req, err := http.NewRequest(method, url.String(), body)
+	if err != nil {
+		return nil, false, err
+	}
+
+	var token string
+	if n.conf.APIToken != "" {
+		token, err = tmplTextFunc(string(n.conf.APIToken))
```

Comment:
why have a template here?

Reply:
Since I'm familiar with OpsGenie, I start the integration based on OpsGenie integration first. The OpsGenie integration also does tmpl on the api key.

https://github.com/prometheus/alertmanager/blob/ab2248afa7e53b78c63e577a10649e6ae8b9db2a/notify/opsgenie/opsgenie.go#L277-L286

I'm just replicate the existing logic without having a deeper knowledge, if it would be necessary.

Reply:
I find personally that templating for credentials is a mistake and brings more problems than value. I would abstain from replicating this "functionality" across receivers.
See also my other comment about leveraging http_config and the fact that dedicated token fields aren't really needed.